samples/core/enumopencl/main.c

```diff
@@ -0,0 +1,243 @@
+    }
+
+    if (errorCode == CL_SUCCESS)
+    {
+        if (size != 0)
+        {
+            str = (char*)malloc(size);
+            if (str == NULL)
+            {
+                errorCode = CL_OUT_OF_HOST_MEMORY;
+            }
+        }
+    }
+
+    if (errorCode == CL_SUCCESS)
+    {
+        errorCode = clGetDeviceInfo(device, param_name, size, str, NULL);
+    }
+
+    if (errorCode != CL_SUCCESS)
+    {
+        free(str);
+    }
+    else
+    {
+        *pStr = str;
+    }
+
+    return errorCode;
+}
+
+static void PrintDeviceType(const char* label, cl_device_type type)
+{
+    printf("%s%s%s%s%s%s\n", label,
+           (type & CL_DEVICE_TYPE_DEFAULT) ? "DEFAULT " : "",
+           (type & CL_DEVICE_TYPE_CPU) ? "CPU " : "",
+           (type & CL_DEVICE_TYPE_GPU) ? "GPU " : "",
+           (type & CL_DEVICE_TYPE_ACCELERATOR) ? "ACCELERATOR " : "",
+           (type & CL_DEVICE_TYPE_CUSTOM) ? "CUSTOM " : "");
+}
+
+static cl_int PrintDeviceInfoSummary(cl_device_id* devices, cl_uint numDevices)
+{
+    cl_int errorCode = CL_SUCCESS;
+
+    cl_device_type deviceType;
+    char* deviceName = NULL;
+    char* deviceVendor = NULL;
+    char* deviceVersion = NULL;
+    char* deviceProfile = NULL;
+    char* driverVersion = NULL;
+
+    cl_uint i = 0;
+    for (i = 0; i < numDevices; i++)
+    {
+        errorCode |= clGetDeviceInfo(devices[i], CL_DEVICE_TYPE,
+                                     sizeof(deviceType), &deviceType, NULL);
+        errorCode |= AllocateAndGetDeviceInfoString(devices[i], CL_DEVICE_NAME,
+                                                    &deviceName);
+        errorCode |= AllocateAndGetDeviceInfoString(
+            devices[i], CL_DEVICE_VENDOR, &deviceVendor);
+        errorCode |= AllocateAndGetDeviceInfoString(
+            devices[i], CL_DEVICE_VERSION, &deviceVersion);
+        errorCode |= AllocateAndGetDeviceInfoString(
+            devices[i], CL_DEVICE_PROFILE, &deviceProfile);
+        errorCode |= AllocateAndGetDeviceInfoString(
+            devices[i], CL_DRIVER_VERSION, &driverVersion);
+
+        if (errorCode == CL_SUCCESS)
+        {
+            printf("Device[%u]:\n", i);
+
+            PrintDeviceType("\tType:           ", deviceType);
+
+            printf("\tName:           %s\n", deviceName);
+            printf("\tVendor:         %s\n", deviceVendor);
+            printf("\tDevice Version: %s\n", deviceVersion);
+            printf("\tDevice Profile: %s\n", deviceProfile);
+            printf("\tDriver Version: %s\n", driverVersion);
+        }
+        else
+        {
+            fprintf(stderr, "Error getting device info for device %u.\n", i);
+        }
+
+        free(deviceName);
+        free(deviceVendor);
+        free(deviceVersion);
+        free(deviceProfile);
+        free(driverVersion);
+
+        deviceName = NULL;
+        deviceVendor = NULL;
+        deviceVersion = NULL;
+        deviceProfile = NULL;
+        driverVersion = NULL;
```

Comment:
Is this idiomatic C to set function local variable to `NULL` after having called `free()` on them?

Reply:
This was a habit I got into to try to avoid possible double-frees, but it's unnecessary for this sample, and I wasn't doing it consistently anyhow.  Removed.